grpc/stream.go

```diff
@@ -0,0 +1,344 @@
+			timer := time.NewTimer(time.Millisecond * 250)
+			select {
+			case <-s.done:
+				// everything is fine
+			case <-timer.C:
+				close(s.done) // hopefully this means we won't double close
+			}
+			timer.Stop()
+		}
+
+		s.tq.Close()
+	}()
+
+	// read & write data from/to the stream
+	go s.readData(readDataChan)
+	go s.writeData()
+
+	ctxDone := ctx.Done()
+	for {
+		select {
+		case msg, ok := <-readDataChan:
+			if !ok {
+				return
+			}
+
+			rt := s.vu.Runtime()
+
+			s.tq.Queue(func() error {
+				listeners := s.eventListeners.all(eventData)
+
+				for _, messageListener := range listeners {
+					if _, err := messageListener(rt.ToValue(msg)); err != nil {
+						// TODO log it?
+						_ = s.closeWithError(err)
+
+						return err
+					}
+				}
+				return nil
+			})
+		case <-ctxDone:
+			// VU is shutting down during an interrupt
+			// stream events will not be forwarded to the VU
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+			ctxDone = nil // this is to block this branch and get through s.done
+		}
+	}
+}
+
+// readData reads data from the stream and forward them to the readDataChan
+func (s *stream) readData(readDataChan chan map[string]interface{}) {
+	defer func() {
+		close(readDataChan)
+	}()
+
+	for {
+		msg, err := s.stream.ReceiveConverted()
+		if errors.Is(err, grpcext.ErrCancelled) {
+			s.logger.Debug("stream is cancelled")
+
+			err = nil
+		}
+
+		if err != nil && !errors.Is(err, io.EOF) {
+			s.logger.WithError(err).Debug("error while reading from the stream")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(err)
+			})
+
+			return
+		}
+
+		// no message means that the stream has been closed
+		if len(msg) == 0 && errors.Is(err, io.EOF) {
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+
+			return
+		}
+
+		select {
+		case readDataChan <- msg:
+		case <-s.done:
+			if len(msg) > 0 {
+				s.logger.Debug("stream is done, sending the last message")
+
+				readDataChan <- msg
+			}
+
+			return
+		}
+	}
+}
+
+// writeData writes data to the stream
+func (s *stream) writeData() {
+	writeChannel := make(chan message)
+	go func() {
+		for {
+			select {
+			case msg, ok := <-writeChannel:
+				if !ok {
+					return
+				}
+
+				err := s.stream.Send(msg.msg)
+				if err != nil {
+					s.logger.WithError(err).Error("failed to send data to the stream")
+
+					s.tq.Queue(func() error {
+						closeErr := s.closeWithError(err)
+						return closeErr
+					})
+					return
+				}
+
+			case <-s.done:
+				return
+			}
+		}
+	}()
+
+	{
+		defer func() {
+			close(writeChannel)
+		}()
+
+		queue := make([]message, 0)
+		var wch chan message
+		var msg message
+		for {
+			wch = nil // this way if nothing to read it will just block
+			if len(queue) > 0 {
+				msg = queue[0]
+				wch = writeChannel
+			}
+			select {
+			case msg = <-s.writeQueueCh:
+				select {
+				case writeChannel <- msg:
+				default:
+					queue = append(queue, msg)
+				}
```

Comment:
Why do we require one more queue? The writeQueueCh should be already enough for it then we could do `s.stream.Send(msg.msg)` directly here.
Otherwise, I think it sounds like a good point for a memory leak in case we can't get a reasonable rate.

Reply:
Like the buffered channel for the writing, this is a copy from the websockets and to be honest, these design with the buffers and queues doesn't look so suspicious to me since they usually that's how we negotiate the async tasks.

I'm not certain about the memory leak without benchmarking the code, but same as the `writeChannel`, this could be a point of a separate investigation.

Reply:
>this is a copy from the websockets and to be honest
>they usually that's how we negotiate the async tasks.

I can't entirely agree, and if there isn't a rationale without it I suggest we refactor the WebSocket code also. I'm not very expert of the new WebSocket codebase so I could be wrong and it might have a specific reason.

If you prefer to investigate it on a specific PR, I'm okay with it.

Reply:
I would argue this likely is the better place for this to stay as it can grow and shrink dynamically.

> memory leak 

We let users write js code - they can already "leak memory" by just keep adding to an array indefinitely. 
I am not against there being some limit, but I have no idea what it should be and what `.send` in websockets (for example) should do. 

Given that the operations are asynchronous I will expect users should know that `send` will not wait and if they keep calling `send` in a loop - it will fill something up somewhere. 

The websocket specification specifically lets you know how many bytes are not written yet. I would argue the gRPC API should have something similar so that users can check if they are not trying to write way more than what is possible. No idea if nodejs, deno or some other gRPC js lib has something we can take inspiration from.